### PR TITLE
Miller-Rabin primality test implementation

### PR DESCRIPTION
**Pull request by tomkrata**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

### CODE REVIEW

**Review by siriak — CHANGES_REQUESTED** [review]
Could you please add some JUnit tests? (see https://github.com/TheAlgorithms/Java/tree/master/src/test/java/com/thealgorithms)

File: src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        long aHi = a >> 24;
+        long aLo = a & ((1 << 24) - 1);
+        long bHi = b >> 24;
+        long bLo = b & ((1 << 24) - 1);
+        long result = ((((aHi * bHi << 16) % m) << 16) % m) << 16;
+        result += ((aLo * bHi + aHi * bLo) << 24) + aLo * bLo;
+        return result % m;
+    }
+
+    public static void main(String[] args) {
```
Comment (siriak) [inline src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java:92]: Please remove main, it's redundant. Tests already test everything

File: src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        long x = power(a, d, n);
+        if (x == 1 || x == n - 1) return false;
+        for (int r = 1; r < s; r++) {
+            x = power(x, 2, n);
+            if (x == n - 1) return false;
+        }
+        return true;
+    }
+
+    private static long power(long x, long y, long p) {
```
Comment (siriak) [inline src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java:64]:
```suggestion
    private static long powerModP(long x, long y, long p) {
```

File: src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            if ((y & 1) == 1) res = moduloMultiplication(res, x, p);
+
+            // y must be even now
+            y = y >> 1; // y = y/2
+            x = moduloMultiplication(x, x, p);
+        }
+        return res;
+    }
+
+    private static long moduloMultiplication(long a, long b, long m) {
```
Comment (siriak) [inline src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java:82]:
```suggestion
    private static long multiplyModP(long a, long b, long p) {
```

File: src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+        for (int a : new int[] {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37}) {
+            if (n == a) return true;
+            if (checkComposite(n, a, d, r)) return false;
+        }
+        return true;
+    }
+
+    private static boolean checkComposite(long n, long a, long d, int s) {
```
Comment (siriak) [inline src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java:54]: Please make parameter names descriptive

File: src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        if (n < 2) return false;
+
+        int r = 0;
+        long d = n - 1;
+        while ((d & 1) == 0) {
+            d >>= 1;
+            r++;
+        }
+
+        for (int a : new int[] {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37}) {
```
Comment (siriak) [inline src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java:47]: We need to test all primes that are below a certain limit as far as I understand

Comment (tomkrata) [reply src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java:47]: I don't really get the math behind the deterministic version, but here is some basic explanation of it https://cp-algorithms.com/algebra/primality_tests.html#deterministic-version.

But from here it shows that it doesn't need to be all primes

Comment (siriak) [reply src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java:47]: From the link you provided:

> It turns out, for testing a 32 bit integer it is only necessary to check the first 4 prime bases: 2, 3, 5 and 7. The smallest composite number that fails this test is  
$3,215,031,751 = 151 \cdot 751 \cdot 28351$ . And for testing 64 bit integer it is enough to check the first 12 prime bases: 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, and 37.

Interesting, so they've found them empirically!

**Review by siriak — APPROVED** [review]
Looks good, thanks!

**Pull request merged**